mobly/base_test.py

```diff
@@ -682,15 +683,22 @@ def generate_tests(self, test_logic, name_func, arg_sets):
                 the test logic function.
             arg_sets: a list of tuples, each tuple is a set of arguments to be
                 passed to the test logic function and name function.
+            uid_func: function, the fu
         """
         self._assert_function_name_in_stack(STAGE_NAME_SETUP_GENERATED_TESTS)
+        root_error_msg = 'Encountered error during test generation of "%s":' % test_logic.__name__
         for args in arg_sets:
             test_name = name_func(*args)
             if test_name in self.get_existing_test_names():
                 raise Error(
-                    'Test name "%s" already exists, cannot be duplicated!' %
-                    test_name)
+                    '%s Test name "%s" already exists, cannot be duplicated!' %
+                    (root_error_msg, test_name))
             test_func = functools.partial(test_logic, *args)
+            if uid_func is not None:
+                uid = uid_func(*args)
+                if uid is None:
+                    raise ValueError('%s UID cannot be None.' % root_error_msg)
+                setattr(test_func, 'uid', uid)
```

Comment:
why not something like to keep the logic consistent
test_func = records.uid(uid)(test_func)

Reply:
`functools` returns a `partial` obj, which is not a straightforward function.
so this impl actually gives me error.
```
base_test.py               275 ERROR    setup_generated_tests failed for MockBaseTest.
Traceback (most recent call last):
  File "/Users/angli/Developer/mobly/mobly/base_test.py", line 272, in _setup_generated_tests
    self.setup_generated_tests()
  File "/Users/angli/Developer/mobly/tests/mobly/base_test_test.py", line 1870, in setup_generated_tests
    arg_sets=[(1, 2), (3, 4)])
  File "/Users/angli/Developer/mobly/mobly/base_test.py", line 704, in generate_tests
    test_func = records.uid(uid)(test_func)
  File "/Users/angli/Developer/mobly/mobly/records.py", line 57, in decorate
    func_name = test_func.__name__
AttributeError: 'functools.partial' object has no attribute '__name__'
```